Add the ability to pass in translations via ptOptions

jonbitgood opened the pull request:
Allows for the localization of the text and numerals via the addition of a new ptOptions.translations object

muonw committed:
.nojekyll

muonw committed:
.nojekyll

muonw committed:
Delete .nojekyll

muonw committed:
Create .nojekyll

muonw committed:
Merge branch 'playground'

muonw committed:
Merge branch 'playground'

muonw committed:
Merge branch 'playground'

muonw committed:
Merge branch 'playground'

muonw committed:
Merge branch 'playground'

muonw committed:
Merge branch 'playground'

muonw committed:
Merge branch 'playground'

muonw committed:
Merge branch 'playground'

jonbitgood committed:
Add a translations object to ptOptions

Allows for the localization of the text and numerals

muonw commented on app/src/lib/components/PowerTable.svelte:
```diff
@@ -141,6 +157,15 @@ let options: Options = {
+        next: 'Next',
+        previous: 'Previous',
+        rows: 'Rows',
+        filterBy: 'Filter by'
```
It's better to set all the default values here and remove them from the HTML code below. That way, for instance `{options.translations?.selectAll ?? 'Select All'}` can be replaced with `{options.translations.selectAll}`

muonw commented on app/src/lib/components/PowerTable.svelte:
```diff
@@ -363,7 +388,7 @@ function applySort() {
         const removeCase = (v: string) => typeof(v) === "string" ? v.toLowerCase() : v;
         const makeCompareFunction = (f:any, opt:any) => {
             opt = typeof(opt) === "object" ? opt : {direction: opt};
-            if(typeof(f)!="function"){var prop = f; f = function(v1:any){return !!v1[prop] ? v1[prop] : "";}}
```
Unclear why this is needed. Can be discussed in a separate pull request.

muonw commented on app/src/lib/components/PowerTable.svelte:
```diff
@@ -433,7 +458,7 @@ function applyFilters() {
 
     if (searchObj.value) {
         // By default search continues after a custom search
-        let customSearchContinue: boolean = true;
```
Unclear how this helps.

muonw commented on app/src/lib/components/PowerTable.svelte:
```diff
@@ -506,7 +531,7 @@ function applyFilters() {
 
         if (filter.value?.length) {
             // By default filter continues after a custom filter
-            let customFilterContinue: boolean = true;
```
Unclear how this helps.

muonw commented on app/src/lib/components/PowerTable.svelte:
```diff
@@ -840,15 +865,15 @@ onMount(async () => {
                 {:else if segment_code.toLowerCase() === 'stats'}
                     <div data-name="stats-container" data-segment_index={segment_index}>
-                        {pagination.firstShownRow}-{pagination.lastShownRow} of {options.filteredRows ?? sortedData.length} 
+                        {options.translations.numeralFormatter.format(pagination.firstShownRow)}-{options.translations.numeralFormatter.format(pagination.lastShownRow)} {options.translations?.of ?? 'of'} {options.translations.numeralFormatter.format(options.filteredRows ?? sortedData.length)} 
```
For all instances of `options.translations.numeralFormatter.format` Typescript complains that 

> 'options.translations.numeralFormatter' is possibly 'undefined'

Using optional chaining and nullish coalescing can suppress the error:

`{options?.translations?.numeralFormatter?.format(pagination.firstShownRow ?? 1)}`

muonw commented on app/src/lib/components/PowerTable.svelte:
```diff
@@ -37,6 +37,22 @@ export interface Options {
     currentPage?: number,
+    translations?: {
+        locale?: string,
+        numeralFormatter?: object,
```
The proper type is probably `Intl.NumberFormat`

muonw commented on app/src/lib/components/PowerTable.svelte:
```diff
@@ -1007,16 +1032,16 @@ onMount(async () => {
                 {:else if segment_code.toLowerCase() === 'settings'}
                     <div data-name="settings-container" data-segment_index={segment_index}>
-                        <button data-name="handle" on:click={toggleMenu}>🛠️</button>
+                        <button data-name="handle" on:click={toggleMenu}>{options.translations?.settings ?? '🛠️'}</button>
```
`settings` is not defined in `translations`

muonw commented on app/src/lib/components/PowerTable.svelte:
```diff
@@ -899,7 +924,7 @@ onMount(async () => {
                                                 <th data-key={instruct.key}>
                                                     {#if instruct?.filterable !== false}
-                                                        <input data-key={instruct.key} data-is_regex={filterObj[instruct.key].isRegex} data-is_custom={filterObj[instruct.key].isCustom} placeholder="Filter by {instruct.title}" bind:value={filterObj[instruct.key].value} on:input={renderTable} >
+                                                        <input data-key={instruct.key} data-is_regex={filterObj[instruct.key].isRegex} data-is_custom={filterObj[instruct.key].isCustom} placeholder="{options.translations?.filterBy ?? 'Filter by'} {instruct.title}" bind:value={filterObj[instruct.key].value} on:input={renderTable} >
```
Another instance of "Filter by" is inside `<tfoot>` that also needs to be updated

muonw commented on app/src/routes/examples/+layout.svelte:
```diff
@@ -89,6 +89,7 @@ let pages = {
             <b>1.</b> <b>Search</b> can only be done on rows that have an <b>even score number</b>. That is because a user-defined function removes all rows that have odd scores (e.g. <code>1</code>,<code>3</code>,<code>5</code>,<code>7</code>) and then returns the remaining data along with a flag that allows the default search function to perform the actual search.<br>
             <b>2.</b> The <b>Score</b> column can be filtered by a <b>number range</b> (two integers separated by a hyphen). When the user-defined function detects a valid range, it filters out all non-matching rows and then returns the filtered data along with a flag that marks the filtering as finalized. If the process is not finalized, the default filter function will look for score numbers th ...
             <b>3.</b> The <b>Birthday</b> column can be filtered by a <b>date range</b>. The methodology is similar to that used for the <b>Score</b> column.<br>
+            <b>4. Add custom localization and numerals via ptOptions.translations</b>
```
Example 3 ([Custom Options](https://muonw.github.io/muonw-powertable/examples/example3)) is more suitable for demonstration. There are two tables on that page and you can apply the translation to either one.

muonw reviewed (changes requested):
Thank you so much for your contribution. This looks fantastic!
I added some comments to suggest a few minor adjustments.

jonbitgood committed:
Moved Translations Example to number 3

- Added Checkbox translation passthrough
- Removed null coalescing from values with defaults
- Updated numeralFormatter Type

muonw reviewed (approved):
Thank you!
